source/fuzz/transformation_split_block.cpp

```diff
@@ -0,0 +1,180 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <utility>
+
+#include "source/fuzz/fuzzer_util.h"
+#include "source/fuzz/transformation_split_block.h"
+#include "source/util/make_unique.h"
+
+namespace spvtools {
+namespace fuzz {
+namespace transformation {
+
+using opt::BasicBlock;
+using opt::IRContext;
+using opt::Instruction;
+using opt::Operand;
+
+namespace {
+
+// Returns:
+// - (true, block->end()) if the relevant instruction is in this block
+//      but inapplicable
+// - (true, it) if 'it' is an iterator for the relevant instruction
+// - (false, _) otherwise.
+std::pair<bool, BasicBlock::iterator> FindInstToSplitBefore(
+    const protobufs::TransformationSplitBlock& message, BasicBlock* block) {
+  // There are three possibilities:
+  // (1) the transformation wants to split at some offset from the block's
+  //     label.
+  // (2) the transformation wants to split at some offset from a
+  //     non-label instruction inside the block.
+  // (3) the split associated with this transformation has nothing to do with
+  //     this block
+  if (message.result_id() == block->id()) {
+    // Case (1).
+    if (message.offset() == 0) {
+      // The offset is not allowed to be 0: this would mean splitting before the
+      // block's label.
+      // By returning (true, block->end()), we indicate that we did find the
+      // instruction (so that it is not worth searching further for it), but
+      // that splitting will not be possible.
+      return {true, block->end()};
+    }
+    // Conceptually, the first instruction in the block is [label + 1].
+    // We thus start from 1 when applying the offset.
+    auto inst_it = block->begin();
+    for (uint32_t i = 1; i < message.offset() && inst_it != block->end(); i++) {
+      ++inst_it;
+    }
+    // This is either the desired instruction, or the end of the block.
+    return {true, inst_it};
+  }
+  for (auto inst_it = block->begin(); inst_it != block->end(); ++inst_it) {
+    if (message.result_id() == inst_it->result_id()) {
+      // Case (2): we have found the base instruction; we now apply the offset.
+      for (uint32_t i = 0; i < message.offset() && inst_it != block->end();
+           i++) {
+        ++inst_it;
+      }
+      // This is either the desired instruction, or the end of the block.
+      return {true, inst_it};
+    }
+  }
+  // Case (3).
+  return {false, block->end()};
+}
+
+}  // namespace
+
+bool IsApplicable(const protobufs::TransformationSplitBlock& message,
```

Comment:
We do expect to use it in places where the basic block isn't available.  In particular, during test case reduction it might be that we remove some transformations that added particular blocks to the module.  We might then still try to apply transformations that split those blocks, and the idea is that IsApplicable would just return false for those transformations (because there would be no blocks to split anymore).